ddtrace/internal/atexit.py

```diff
@@ -0,0 +1,59 @@
+# -*- encoding: utf-8 -*-
+"""
+An API to provide atexit functionalities
+"""
+import atexit
+import logging
+import typing
+
+
+log = logging.getLogger(__name__)
+
+
+if hasattr(atexit, "unregister"):
+    register = atexit.register
+    unregister = atexit.unregister
+else:
+    # Hello Python 2!
+    _registry = []  # type: typing.List[typing.Callable[[], None]]
+
+    def _ddtrace_atexit():
+        # type: (...) -> None
+        """Wrapper function that calls all registered function on normal program termination"""
+        global _registry
+
+        # DEV: we make a copy of the registry to prevent hook execution from
+        # introducing new hooks, potentially causing an infinite loop.
+        for hook in list(_registry):
+            try:
+                hook()
+            except Exception:
+                # Mimic the behaviour of Python's atexit hooks.
+                log.exception("Error in atexit hook %r", hook)
+
+    atexit.register(_ddtrace_atexit)
+
+    def register(
+        func,  # type: typing.Callable[..., typing.Any]
+        *args,  # type: typing.Any
+        **kwargs,  # type: typing.Any
+    ):
+        # type: (...) -> typing.Callable[..., typing.Any]
+        """Register a function to be executed upon normal program termination"""
+
+        def _w():
+            func(*args, **kwargs)
+
+        _registry.append(_w)
+        return func
+
+    def unregister(func):
+        # type: (typing.Callable[..., None]) -> None
+        """
+        Unregister an exit function which was previously registered using
+        atexit.register
```

Comment:
Worth pointing out the behaviour when a function that isn't registered is unregistered?

Reply:
For the anecdote, I copied Python's `atexit` docstring.